lib/srv/server/ssm_install.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ssm"
+	"github.com/aws/aws-sdk-go/service/ssm/ssmiface"
+	"github.com/gravitational/teleport/api/types/events"
+	libevent "github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/trace"
+)
+
+type SSMInstaller struct {
+	instances    []*string
+	SSM          ssmiface.SSMAPI
+	recheckTimer *time.Ticker
+	params       map[string][]*string
+	emitter      events.Emitter
+	ctx          context.Context
+	cancelfn     context.CancelFunc
+	region       string
+	accountID    string
+}
+
+// SSMInstallerConfig represents configuration for an SSM install
+// script executor.
+type SSMInstallerConfig struct {
+	// SSM is an SSM API client.
+	SSM ssmiface.SSMAPI
+	// Instances is the list of instances that will have the SSM
+	// document executed on them.
+	Instances []*ec2.Instance
+	// Params is a list of parameters to include when executing the
+	// SSM document.
+	Params map[string]string
+	// Emitter is an events emitter.
+	Emitter events.Emitter
+	Ctx     context.Context
+	// Region is the region instances are present in, used in audit
+	// events.
+	Region string
+	// AccountID is the AWS account being used to execute the SSM document.
+	AccountID string
+}
+
+func NewSSMInstaller(cfg SSMInstallerConfig) *SSMInstaller {
+	var ids []*string
+
+	for _, inst := range cfg.Instances {
+		ids = append(ids, inst.InstanceId)
+	}
+	ssmParams := make(map[string][]*string)
+
+	for key, val := range cfg.Params {
+		ssmParams[key] = []*string{aws.String(val)}
+	}
+
+	return &SSMInstaller{
+		instances:    ids,
+		SSM:          cfg.SSM,
+		recheckTimer: time.NewTicker(time.Second * 30),
```

Comment:
If you implement my suggestion above with structs reorg, this ticket will just move to Run method.